tests/keras/engine/test_training.py

```diff
@@ -1173,5 +1173,39 @@ def test_pandas_dataframe():
                           [output_a_df, output_b_df])
 
 
+@keras_test
+def test_model_with_sparse_loss_channels_first():
+    """
+    Tests correctness of _standardize_user_data when using sparse_categorical_crossentropy loss with 'channels_first'
+    image_data_format.
+
+    Verifies that error no longer occurs after PR #9715.
+    """
+    old_data_format = K.image_data_format()
+
+    K.set_image_data_format('channels_last')
+
+    data = np.zeros((1, 3, 3, 1))
+    labels = np.array([[[[0, 1, 0], [2, 1, 0], [2, 2, 1]]]])
+    labels = np.moveaxis(labels, 1, -1)
+    inputs = Input(shape=(3, 3, 1))
+    predictions = Conv2D(3, 1, activation='softmax')(inputs)
+    model = Model(inputs=inputs, outputs=predictions)
+    model.compile(optimizer='rmsprop', loss='sparse_categorical_crossentropy')
+    model.fit(data, labels)
+
+    K.set_image_data_format('channels_first')
+
+    data = np.zeros((1, 1, 3, 3))
+    labels = np.array([[[[0, 1, 0], [2, 1, 0], [2, 2, 1]]]])
+    inputs = Input(shape=(1, 3, 3))
+    predictions = Conv2D(3, 1, activation='softmax')(inputs)
+    model = Model(inputs=inputs, outputs=predictions)
+    model.compile(optimizer='rmsprop', loss='sparse_categorical_crossentropy')
+    model.fit(data, labels)
```

Comment:
You don't appear to be doing any correctness checking. A simpler test would be:

- call `predict` with the first model, store results in y1
- switch the data format, call `predict` on the same input array, store results in y2
- check that y1 == y2

Reply:
@fchollet The bug fix I am testing is in `_standardize_user_data`, which is called (among other places) from the `fit` function. When I run this test without the bug fix, with `'channels_first'`, I get the exception ```"ValueError: Error when checking target: expected conv2d_2 to have shape (3, 3, 1) but got array with shape (1, 3, 3)"```, because [this line of code](https://github.com/keras-team/keras/blob/068a680480ebf27ecbe57f1406d60157b7d2df38/keras/engine/training.py#L1461) didn't take into account the possibility of  `'channels_first'` data format. After the bug fix no exception is raised and the test passes.

Reply:
Got it. In that case, prefer using `evaluate` and check that the resulting losses are the same (on the same data) with both data formats.

Reply:
When I do that I run into additional places that assume `'channels_last'`, such as [this line](https://github.com/keras-team/keras/blob/be112e1f48bbdf24d0a714b9cafc3fc9f43c7738/keras/backend/tensorflow_backend.py#L3041) and [this line](https://github.com/keras-team/keras/blob/be112e1f48bbdf24d0a714b9cafc3fc9f43c7738/keras/backend/tensorflow_backend.py#L3048) in `tensorflow_backend` `sparse_categorical_crossentropy`. In general `'channels_first'` does not seem to be supported in many places in the codebase, for which I opened issue #9670. This pull request fixes one such issue, but not all.

Reply:
The best would be to add an `axis` argument in these loss functions.

If you don't want to do that, you can use a different loss function, like `mse`.

Reply:
@fchollet Why add an `axis` argument? The code sits in `tensorflow_backend`, so why shouldn't it just include a check "`if image_data_format() == 'channels_last':`"?